Project 2/k-NNAlgorithms/src/Algorithms.java

```diff
@@ -0,0 +1,171 @@
+            prevRecall=currentRecall+.05;
+
+            currentPrecision= Double.parseDouble(results.get(0));
+            currentRecall = Double.parseDouble(results.get(1));
+            currentAccuracy= Double.parseDouble(results.get(2));
+
+        }
+        //return the set of edited training data just prior to run that had a decrease
+        //in precision, accuracy or recall
+        return  KNN(finalEditedTrainingData,testingData, k, regression, euclidean);
+    }
+
+
+
+    public static ArrayList<String> CondensedKNN(ArrayList<ArrayList<String>>trainingData, ArrayList<ArrayList<String>>testingData, int k, boolean regression, boolean euclidean){
+        //initialize empty set
+        ArrayList<ArrayList<String>> condensedTrainingData=new ArrayList<>();
+        int currSizeOfCondensed=-1;
+        int indexOfClassification=trainingData.get(0).size()-1;
+        int lengthOfFeatures = trainingData.get(0).size()-2;
+        //while  condensedTrainingData size is increasing
+        while(condensedTrainingData.size()>currSizeOfCondensed){
+            //randomize order
+            currSizeOfCondensed=condensedTrainingData.size();
+            Collections.shuffle(trainingData);
+            for (int i = 0; i <trainingData.size() ; i++) {
+                //if condensed has no elements, add the first
+                if(condensedTrainingData.size()==0){
+                    condensedTrainingData.add(trainingData.get(i));
+                }else{
+                    //find minimum distance between current point and points in the condensed set
+                    ArrayList<Double> distanceToAllPoints = new ArrayList<Double>();
+                    for (int j = 0; j <condensedTrainingData.size() ; j++) {
+
+                        List<String> condensedTrainingFeatures = condensedTrainingData.get(j).subList(0,lengthOfFeatures);
+                        List<String> trainingFeatures = trainingData.get(i).subList(0,lengthOfFeatures);
+                        if(euclidean){
+                            distanceToAllPoints.add(MathFunction.euclideanDistance(condensedTrainingFeatures, trainingFeatures));
+                        }else{
+                            distanceToAllPoints.add(MathFunction.hammingDistance(condensedTrainingFeatures, trainingFeatures));
+                        }
+
+                    }
+                    int indexOfMinValue= distanceToAllPoints.indexOf(Collections.min(distanceToAllPoints));
+                    String classOfMinCondensedPoint = condensedTrainingData.get(indexOfMinValue).get(indexOfClassification);
+
+                    //add the point to condensed training data if it's classification doesn't equal the classification of the point that
+                    //is min distance from it, and it doesn't already exist in condensed
+                    if(condensedTrainingData.get(indexOfMinValue).get(indexOfClassification).equals(trainingData.get(i).get(indexOfClassification))){
```

Comment:
This logic block can probably be rewritten such that we don't need paths that don't execute any code (negate the condition or something). Also another potential candidate for giving a name to the condition since it's so long.

Reply:
I agree, I was being somewhat sloppy this morning